File: src/engines/jianying-adapter/python-worker/zhiheng_jianying_worker/contract.py
# -*- coding: utf-8 -*-
"""Job / Result Contract —— 与 TS 侧 src/engines/jianying-adapter/contract.ts 保持一致。

共享 Contract Fixtures 位于：
  src/engines/jianying-adapter/__fixtures__/
TS 与 Python 两侧读取同一批 JSON 做交叉验证，防止两套 schema 漂移。
"""

import json
import os
from pathlib import Path

from .errors import ERROR_CODE_SET, WorkerError

CONTRACT_VERSION = "0.1.0"
SUPPORTED_TIMELINE_SCHEMA_VERSION = 2

# PJD 版本锁定元数据（Phase C.1 / C.2 / 高版本 fork 接入）
# 高版本 fork：aoguai/pyJianYingDraft @ main 4a7730c9（基于上游 GuanYixuan c3318066 后 4 个提交）
# - 本地 git 仓库：D:\剪映智剪测试\pyJianYingDraft-fork-v0（zip snapshot，无 upstream git 历史）
# - 本地锁定 commit = fdd9c04f...（git init 生成的本地 commit，上游真实 commit = 4a7730c9）
# - 已验证：fork 生成明文草稿可被剪映 11.3.0.14362 首次正常打开（关键：root_meta_info 注册 + meta 完整）
PJD_REPOSITORY = "aoguai/pyJianYingDraft"
PJD_EXPECTED_COMMIT = "fdd9c04fd44257222aa1af45fdd7c4ac029e652e"  # 本地锁定 commit（全量比较）
PJD_UPSTREAM_COMMIT = "4a7730c9a14e91aa497e723c85b5c433a62a163c"  # 上游真实 commit（参考记录）
PJD_PACKAGE_VERSION = "fork-v0@4a7730c9"
PJD_ROOT_ENV = "ZHIHENG_PJD_ROOT"  # 统一拼写（不再使用 ZHIJING_PJD_ROOT）

# 共享 fixtures 目录：zhiheng_jianying_worker/contract.py -> ../../__fixtures__
FIXTURES_DIR = Path(__file__).resolve().parent.parent.parent / "__fixtures__"

# 生产 ResourceMap 数据：jianying-adapter/resources/resource-map.v0.json
# （生产代码只读取正式资源文件，不依赖 __fixtures__；与 TS 侧同一文件，真正单源）
PRODUCTION_RESOURCE_MAP_PATH = (
    Path(__file__).resolve().parent.parent.parent / "resources" / "resource-map.v0.json"
)

SHARED_FIXTURE_FILES = (
    "job-minimal.json",
    "job-dissolve.json",
    "job-keyword.json",
    "result-ok.json",
    "result-error.json",
    "error-codes.json",
)


def load_production_resource_map():
    """读取生产 ResourceMap 数据（resources/resource-map.v0.json）。"""
    if not PRODUCTION_RESOURCE_MAP_PATH.exists():
        raise WorkerError(
            "JOB_INVALID",
            "生产 ResourceMap 数据不存在: %s" % PRODUCTION_RESOURCE_MAP_PATH,
        )
    with open(PRODUCTION_RESOURCE_MAP_PATH, encoding="utf-8") as f:
        return json.load(f)


def load_fixture(name):
    """读取共享 fixture JSON（TS/Python 交叉验证用）。"""
    p = FIXTURES_DIR / name
    if not p.exists():
        raise WorkerError("JOB_INVALID", "共享 fixture 不存在: %s" % p)
    with open(p, encoding="utf-8") as f:
        return json.load(f)


def validate_job(job):
    """校验 Job 结构（最小但关键字段）。

    完整结构由 TS 侧 zod 校验；Python 侧负责契约级检查：
    - contractVersion / timelineSchemaVersion 匹配
    - 关键字段存在且类型正确
    失败抛 WorkerError。
    """
    if not isinstance(job, dict):
        raise WorkerError("JOB_INVALID", "Job 必须是 JSON 对象")
    if job.get("contractVersion") != CONTRACT_VERSION:
        raise WorkerError(
            "UNSUPPORTED_CONTRACT_VERSION",
            "不支持的 contractVersion: %r（期望 %s）" % (job.get("contractVersion"), CONTRACT_VERSION),
        )
    if job.get("timelineSchemaVersion") != SUPPORTED_TIMELINE_SCHEMA_VERSION:
        raise WorkerError(
            "UNSUPPORTED_TIMELINE_VERSION",
            "不支持的 timelineSchemaVersion: %r（期望 %s）"
            % (job.get("timelineSchemaVersion"), SUPPORTED_TIMELINE_SCHEMA_VERSION),
        )
    job_id = job.get("jobId")
    if not isinstance(job_id, str) or not job_id:
        raise WorkerError("JOB_INVALID", "jobId 缺失或非法")
    for key in ("draft", "timeline", "assetRoot", "outputDraftDir", "stagingRoot", "logDir", "options"):
        if key not in job:
            raise WorkerError("JOB_INVALID", "Job 缺少字段: %s" % key)
    options = job.get("options", {})
    backup_plaintext = bool(options.get("backupPlaintext", False))
    if backup_plaintext:
        # Phase C.2：backupPlaintext=true 时 backupRoot 必填
        if not isinstance(job.get("backupRoot"), str) or not job.get("backupRoot"):
            raise WorkerError("JOB_INVALID", "backupPlaintext=true 时 backupRoot 必填")
    draft = job["draft"]
    for key in ("name", "width", "height", "fps"):
        if key not in draft:
            raise WorkerError("JOB_INVALID", "draft 缺少字段: %s" % key)
    timeline = job["timeline"]
    if not isinstance(timeline, dict) or "videoTrack" not in timeline:
        raise WorkerError("JOB_INVALID", "timeline 缺少 videoTrack")
    return job_id


def validate_result(result):
    """校验 Result 结构（Worker 输出前自检 + 测试用）。失败抛 WorkerError。"""
    if not isinstance(result, dict):
        raise WorkerError("WORKER_PROTOCOL_ERROR", "Result 必须是 JSON 对象")
    if result.get("contractVersion") != CONTRACT_VERSION:
        raise WorkerError("WORKER_PROTOCOL_ERROR", "Result contractVersion 不匹配")
    if "ok" not in result or "jobId" not in result:
        raise WorkerError("WORKER_PROTOCOL_ERROR", "Result 缺少 ok/jobId")
    if result.get("ok"):
        for key in ("draftDir", "duration", "tracks", "validationReport"):
            if key not in result:
                raise WorkerError("WORKER_PROTOCOL_ERROR", "Result 缺少字段: %s" % key)
    else:
        err = result.get("error")
        if not err or err.get("code") not in ERROR_CODE_SET:
            raise WorkerError("WORKER_PROTOCOL_ERROR", "Result error 缺失或非法")
    return result


def cross_check_error_codes_fixture():
    """交叉验证 error-codes.json fixture 与本地错误码集合一致。"""
    fixture = load_fixture("error-codes.json")
    codes = fixture.get("errorCodes")
    if not isinstance(codes, list):
        raise WorkerError("WORKER_PROTOCOL_ERROR", "error-codes.json 格式非法")
    if set(codes) != ERROR_CODE_SET:
        raise WorkerError(
            "WORKER_PROTOCOL_ERROR",
            "error-codes fixture 与本地错误码不一致: fixture=%s local=%s"
            % (sorted(codes), sorted(ERROR_CODE_SET)),
        )
    return codes

File: src/engines/jianying-adapter/contract.ts
/**
 * JianYing Adapter —— Contract 解析与共享 Fixtures。
 *
 * Job/Result 用 zod 校验，保证 TS 侧结构完整性。
 * 共享 Contract Fixtures 放在 __fixtures__/ 下，TS 与 Python Worker 交叉验证用。
 */
import { z } from 'zod';
import fs from 'node:fs';
import path from 'node:path';
import { UnifiedTimelineV2Schema } from '../zhiheng-renderer/v2-types';
import { ErrorCode, ERROR_CODE_LIST, type ContractError } from './errors';
// 使用 namespace import + 显式取值，避免 esbuild/tsx 将命名导入值绑定转为 undefined
import * as contractTypes from './types';
import type { JianYingJob, JianYingResult } from './types';

// 常量统一来自 types.ts（单一事实源），这里显式取值并 re-export
export const CONTRACT_VERSION = contractTypes.CONTRACT_VERSION;
export const SUPPORTED_TIMELINE_SCHEMA_VERSION = contractTypes.SUPPORTED_TIMELINE_SCHEMA_VERSION;
export const WORKER_PYTHON_MODULE = contractTypes.WORKER_PYTHON_MODULE;

// ============================================================================
// Job zod schema
// ============================================================================

const DraftSpecSchema = z.object({
  name: z.string().min(1),
  width: z.number().int().positive(),
  height: z.number().int().positive(),
  fps: z.number().positive()
});

const JobOptionsSchema = z.object({
  backupPlaintext: z.boolean().default(false),
  failOnWarning: z.boolean().default(false)
});

/** Job Contract 校验 schema */
export const JianYingJobSchema = z
  .object({
    contractVersion: z.literal(CONTRACT_VERSION),
    jobId: z.string().min(1),
    timelineSchemaVersion: z.literal(SUPPORTED_TIMELINE_SCHEMA_VERSION),
    draft: DraftSpecSchema,
    timeline: UnifiedTimelineV2Schema,
    resourceMapRef: z.string().min(1),
    assetRoot: z.string().min(1),
    outputDraftDir: z.string().min(1),
    stagingRoot: z.string().min(1),
    backupRoot: z.string().min(1).optional(),
    officialDraftRoot: z.string().min(1).optional(),
    logDir: z.string().min(1),
    options: JobOptionsSchema
  })
  // Phase C.2：backupPlaintext=true 时 backupRoot 必填
  .refine((j) => !j.options.backupPlaintext || !!j.backupRoot, {
    path: ['backupRoot'],
    message: 'backupPlaintext=true 时 backupRoot 必填'
  });

/** Result Contract 校验 schema */
export const JianYingResultSchema = z.object({
  contractVersion: z.literal(CONTRACT_VERSION),
  jobId: z.string().min(1),
  ok: z.boolean(),
  draftDir: z.string().optional(),
  duration: z.number().optional(),
  tracks: z.array(z.object({ type: z.string(), count: z.number() })).optional(),
  warnings: z.array(z.string()).default([]),
  manualReviewRequired: z.boolean().default(false),
  validationReport: z.object({
    fileCount: z.number(),
    hasDraftContent: z.boolean(),
    hasDraftMetaInfo: z.boolean(),
    hasDraftInfo: z.boolean(),
    duration: z.number(),
    passed: z.boolean(),
    checks: z.array(z.object({ name: z.string(), ok: z.boolean(), detail: z.string().optional() })),
    pjdCommit: z.string().optional(),
    pjdVersion: z.string().optional(),
    pjdSource: z
      .object({
        expectedCommit: z.string(),
        actualCommit: z.string(),
        repositoryRemote: z.string(),
        moduleFile: z.string(),
        sourceDirty: z.boolean(),
        packageVersion: z.string(),
        pythonVersion: z.string()
      })
      .optional(),
    logFile: z.string().optional(),
    backupManifest: z.string().optional()
  }),
  error: z
    .object({
      code: z.enum(ERROR_CODE_LIST as [string, ...string[]]),
      message: z.string()
    })
    .optional()
});

/** 解析并校验 Job JSON（返回错误码而非抛未知异常） */
export function parseJob(input: unknown): { job: JianYingJob } | { error: ContractError } {
  if (typeof input !== 'object' || input === null) {
    return { error: { code: ErrorCode.JOB_INVALID, message: 'Job 必须是 JSON 对象' } };
  }
  const raw = input as Record<string, unknown>;
  // 版本错误优先于结构错误（仅当版本字段存在但不被支持时；缺失字段归 JOB_INVALID）
  if (raw.contractVersion !== undefined && raw.contractVersion !== CONTRACT_VERSION) {
    return {
      error: {
        code: ErrorCode.UNSUPPORTED_CONTRACT_VERSION,
        message: `不支持的 contractVersion: ${String(raw.contractVersion)}（期望 ${CONTRACT_VERSION}）`
      }
    };
  }
  if (
    raw.timelineSchemaVersion !== undefined &&
    raw.timelineSchemaVersion !== SUPPORTED_TIMELINE_SCHEMA_VERSION
  ) {
    return {
      error: {
        code: ErrorCode.UNSUPPORTED_TIMELINE_VERSION,
        message: `不支持的 timelineSchemaVersion: ${String(raw.timelineSchemaVersion)}（期望 ${SUPPORTED_TIMELINE_SCHEMA_VERSION}）`
      }
    };
  }
  const parsed = JianYingJobSchema.safeParse(input);
  if (!parsed.success) {
    const first = parsed.error.issues[0];
    return {
      error: {
        code: ErrorCode.JOB_INVALID,
        message: `Job 无效：${first ? first.path.join('.') + ': ' + first.message : '未知结构错误'}`
      }
    };
  }
  return { job: parsed.data as JianYingJob };
}

/** 解析并校验 Result JSON（Worker 输出） */
export function parseResult(input: unknown): { result: JianYingResult } | { error: ContractError } {
  const parsed = JianYingResultSchema.safeParse(input);
  if (!parsed.success) {
    const first = parsed.error.issues[0];
    return {
      error: {
        code: ErrorCode.WORKER_PROTOCOL_ERROR,
        message: `Result 无效：${first ? first.path.join('.') + ': ' + first.message : '未知结构错误'}`
      }
    };
  }
  return { result: parsed.data as JianYingResult };
}

// ============================================================================
// 共享 Contract Fixtures（TS 与 Python 交叉验证）
// ============================================================================

/** __fixtures__ 目录绝对路径（Next.js 打包下 __dirname 为虚拟 \ROOT\，回退到项目相对路径） */
export function fixturesDir(): string {
  const viaDirname = path.join(__dirname, '__fixtures__');
  if (fs.existsSync(viaDirname)) return viaDirname;
  return path.join(process.cwd(), 'src', 'engines', 'jianying-adapter', '__fixtures__');
}

/** 读取共享 fixture 文件 */
export function loadFixture<T = unknown>(name: string): T {
  const p = path.join(fixturesDir(), name);
  if (!fs.existsSync(p)) {
    throw new Error(`Fixture 不存在: ${p}`);
  }
  return JSON.parse(fs.readFileSync(p, 'utf-8')) as T;
}

/** fixture 文件名列表（供 Python 侧对照） */
export const SHARED_FIXTURE_FILES = [
  'job-minimal.json',
  'job-dissolve.json',
  'job-keyword.json',
  'result-ok.json',
  'result-error.json',
  'error-codes.json'
] as const;
